src/delay.ml

```diff
@@ -307,6 +307,74 @@ let parse_durationl l =
   let s = String.concat " " l in
   parse_duration s
 
+let duration_of_daypack_duration (x : Daypack_lib.Duration.t) : duration =
+  C.Period.lmake ~day:x.days ~hour:x.hours ~minute:x.minutes ~second:x.seconds ()
+
+let get_current_tz_offset_s () =
+  Ptime_clock.current_tz_offset_s () |> Option.get |> Option.some
+
+let search_param =
+  let open Daypack_lib in
+  Search_param.Years_ahead_start_unix_second {
+    search_using_tz_offset_s = get_current_tz_offset_s ();
+    start = Time.Current.cur_unix_second ();
+    search_years_ahead = 5;
+  }
+
+let parse_durationl' l =
+  let s = String.concat " " l in
+  Daypack_lib.Duration.of_string s
+  |> Result.map duration_of_daypack_duration
+
+let parse_time_pointl' l =
+  let open Daypack_lib in
+  let s = String.concat " " l in
+  match Time_expr.of_string s with
+  | Ok x ->
+    (match Time_expr.next_match_time_slot search_param
+             x with
```

Comment:
If I understand the code correctly, this does a search on the next 5 years for a time slot that correspond to the specification.

Isn't there a way to just say "Does this time expr correspond to a unique point in time, and give it to me" ? This seems like it should be in your API.

Reply:
In particular, this feels important: if the user provides a notification time that is non-unique, I think we should fail.

Reply:
> If I understand the code correctly, this does a search on the next 5 years for a time slot that correspond to the specification.

That is correct. (0.0.4 handles the case differently if the user also provides the years, namely daypack extends the search range to include the years if they were originally outside of the search range.)

> Isn't there a way to just say "Does this time expr correspond to a unique point in time, and give it to me" ? This seems like it should be in your API.

One problem would be time expr right now always reduces to lazy sequences of all possible matches, e.g. "fri 10am" would give you a `int64 Seq.t` of all friday 10ams.

There was originally an idea of bounding the sequences (so you'd have keywords like "next" or "every", where "next" is close to what you meant), but it became too difficult for me to nail down an intuitive set of semantics especially after the introduction of set operators.

More specifically, I couldn't figure out the appropriate meaning for cases like

```
(5pm || 6pm || 8pm)
&& (jun 10 00:00 to jun 10 10am || july 8pm to 11:45pm)
```

Should 5pm be "next" by default? or "every"? What about 6pm? etc

The semantics seemed easier to define if they are just all lazy sequences/sets that span the search range (could be just me being silly as well).

> In particular, this feels important: if the user provides a notification time that is non-unique, I think we should fail as it starts falling out of scope of the simple fugit use-case, and we can't be sure what they mean.

Overall this is closer to insufficiency of the language/library than user being unclear.

There was a point where it was possible to use only a fragment of time expr, such that time points are different from time slots, but then allowing them to mix using set operators became difficult to reason with and implement and I gave up the divide.

But it seems like it might still be worthwhile to allow restriction on the fragment of the language to use? Especially since I can't tell if fugit would find set operations useful.

Reply:
So far, I explicitly do *not* want Fugit to become a cron-like, so set operations are excluded and the meaning is always "next".

Reply:
As for the more general question. It seems to me that your current semantics is always `every` and `next` could be defined as an operator that cut the sequence it is given after the first element.

Reply:
(That being said, if I eventually get around of adding a mode to insert into my calendar for fugit, I wouldn't mind having `every` for that ... but only for that mode, and it would be explicit)

Reply:
> So far, I explicitly do not want Fugit to become a cron-like, so set operations are excluded and the meaning is always "next".

Okay yep, I think separating out at least the time point fragment is worthwhile then. Back to the drawing board~

> As for the more general question. It seems to me that your current semantics is always every and next could be defined as an operator that cut the sequence it is given after the first element.

There were some non-trivial cases for meaning of "next", where defining "next" directly over the underlying sequence may violate the intuitive understanding of the specification.

So things like `jun . 10 to 20 . 1pm to 2pm`, it would seem intuitive that "next" mean the "next batch" of the jun 10th to 20th, 1pm to 2pm, rather than just "jun 10th 1pm to 2pm". Similar confusion carries over for other implicitly branching clauses.

Granted those are problems specific to the time slots expr, and (i think) don't plague time point.

> (That being said, if I eventually get around of adding a mode to insert into my calendar with fugit, I wouldn't mind having every for that ... but only for that mode, and it would be explicit)

That reminds me: daypack may be able to interact with ical files directly someday, with import/export of its native schedule format.

---

I'll be revisiting how to "bound" elegantly over next few days. While it makes sense for how to define them for the case of time points, I would want the operators to also work for the full fragment of time expr (or try to make that so anyway).